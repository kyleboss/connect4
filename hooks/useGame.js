import { useContext } from "react";

import GameContext from "@/state/GameContext";

const useGame = () => useContext(GameContext);

export default useGame;
