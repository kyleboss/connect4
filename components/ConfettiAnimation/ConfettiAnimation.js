import { useEffect, useState, useRef } from "react";
import styled from "styled-components";
import lottie from "lottie-web";

import confetti from "@/animations/confetti.json";
import useGame from "@/hooks/useGame";
import { GAME_STATE_GAME_OVER } from "@/utils/constants";

const AnimationContainer = styled.div`
  width: 100vw;
  height: 100vh;
  position: fixed;
  top: 0;
  left: 0;
  pointer-events: none;
`;

const ConfettiAnimation = () => {
  const animationContainer = useRef(null);
  const { userRole, gameState, winner } = useGame();
  const [showConfetti, setShowConfetti] = useState(false);

  useEffect(() => {
    let confettiTimeout;
    if (gameState === GAME_STATE_GAME_OVER && winner === userRole) {
      setShowConfetti(true);
      confettiTimeout = setTimeout(() => setShowConfetti(false), 5000);
    }

    return () => {
      setShowConfetti(false);
      clearTimeout(confettiTimeout);
    };
  }, [gameState, userRole, winner]);

  useEffect(() => {
    const anim = lottie.loadAnimation({
      container: animationContainer.current,
      renderer: "svg",
      loop: true,
      autoplay: true,
      animationData: confetti,
    });

    if (showConfetti) {
      anim.goToAndPlay(0);
    } else {
      anim.stop();
    }

    return () => anim.destroy();
  }, [showConfetti]);

  return <AnimationContainer ref={animationContainer} />;
};

export default ConfettiAnimation;
