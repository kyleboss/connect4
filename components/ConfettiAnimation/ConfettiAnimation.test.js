import React from "react";
import { render } from "@testing-library/react";
import lottie from "lottie-web";

import ConfettiAnimation from "./ConfettiAnimation";
import useGame from "../../hooks/useGame";

jest.mock("lottie-web", () => ({
  loadAnimation: jest.fn().mockReturnValue({
    goToAndPlay: jest.fn(),
    stop: jest.fn(),
    destroy: jest.fn(),
  }),
}));

jest.mock("../../hooks/useGame", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("ConfettiAnimation", () => {
  it("renders correctly when it is the user's win", () => {
    useGame.mockImplementation(() => ({
      userRole: "Red",
      gameState: "GAME_STATE_GAME_OVER",
      winner: "Red",
    }));

    const { asFragment } = render(<ConfettiAnimation />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly when it is the user's win", () => {
    useGame.mockImplementation(() => ({
      userRole: "Red",
      gameState: "GAME_STATE_GAME_OVER",
      winner: "Red",
    }));

    const { asFragment } = render(<ConfettiAnimation />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("loads but does not play animation when it is not the user's win", () => {
    useGame.mockImplementation(() => ({
      userRole: "Yellow",
      gameState: "GAME_STATE_GAME_OVER",
      winner: "Red",
    }));

    render(<ConfettiAnimation />);
    expect(lottie.loadAnimation).toHaveBeenCalled();
    expect(lottie.loadAnimation().goToAndPlay).not.toHaveBeenCalled();
  });
});
