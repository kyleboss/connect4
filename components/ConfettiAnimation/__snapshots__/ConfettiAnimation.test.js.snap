// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfettiAnimation renders correctly when it is the user's win 1`] = `
<DocumentFragment>
  <div
    class="sc-94caac3c-0 jfkWeT"
  />
</DocumentFragment>
`;

exports[`ConfettiAnimation renders correctly when it is the user's win 2`] = `
<DocumentFragment>
  <div
    class="sc-94caac3c-0 jfkWeT"
  />
</DocumentFragment>
`;
